Check for and use gmake on BSD systems

FreeBSD uses bmake by default. The build script should use ``gmake`` to build libjemalloc if its available. 

This should fix the libjemalloc error.

```
depbase=`echo mapper.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`; gcc -DHAVE_CONFIG_H -I. -I/usr/home/tony/study/python/pyston/build_deps/libunwind/tests -I../include  -I/usr/home/tony/study/python/pyston/build_deps/libunwind/include  -D_GNU_SOURCE -DDEBUG  -O0 -g -fexceptions -Wall -Wsign-compare -MT mapper.o -MD -MP -MF $depbase.Tpo -c -o mapper.o /usr/home/tony/study/python/pyston/build_deps/libunwind/tests/mapper.c && mv -f $depbase.Tpo $depbase.Po
/usr/home/tony/study/python/pyston/build_deps/libunwind/tests/mapper.c: In function 'main':
/usr/home/tony/study/python/pyston/build_deps/libunwind/tests/mapper.c:55:33: error: 'MAP_NORESERVE' undeclared (first use in this function)
   MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE,
                                 ^
/usr/home/tony/study/python/pyston/build_deps/libunwind/tests/mapper.c:55:33: note: each undeclared identifier is reported only once for each function it appears in
*** [mapper.o] Error code 1

make[2]: stopped in /usr/home/tony/study/python/pyston/build/Debug/build_deps/libunwind/src/libunwind-build/tests
1 error

make[2]: stopped in /usr/home/tony/study/python/pyston/build/Debug/build_deps/libunwind/src/libunwind-build/tests
*** [all-recursive] Error code 1

make[1]: stopped in /usr/home/tony/study/python/pyston/build/Debug/build_deps/libunwind/src/libunwind-build

make[1]: stopped in /usr/home/tony/study/python/pyston/build/Debug/build_deps/libunwind/src/libunwind-build
                                                                                                                                                               
~/
❯ cat /usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-stamp/libjemalloc-build-*.log

make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 3: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 4: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 63: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 65: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 66: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 68: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 70: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 71: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 86: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 88: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 89: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 91: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 92: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 94: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 96: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 98: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 100: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 142: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 146: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 190: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 191: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 192: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 193: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 194: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 195: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 196: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 213: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 216: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 217: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 222: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 224: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 226: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 233: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 238: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 240: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 242: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 246: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 293: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 295: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 346: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 374: Need an operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 420: Missing dependency operator
make[1]: "/usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build/Makefile" line 438: Need an operator
make[1]: Fatal errors encountered -- cannot continue
make[1]: stopped in /usr/home/tony/study/python/pyston/build/Debug/jemalloc-build/src/libjemalloc-build
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -21,6 +21,23 @@ if(NOT ${CMAKE_BUILD_TYPE} STREQUAL "Release" AND NOT ${CMAKE_BUILD_TYPE} STREQU
   message(FATAL_ERROR "CMAKE_BUILD_TYPE must be set to Release or Debug")
 endif()
 
+if(UNIX)
+  find_program(MAKE_PRG NAMES gmake make)
+  if(MAKE_PRG)
+    execute_process(
+      COMMAND "${MAKE_PRG}" --version
+      OUTPUT_VARIABLE MAKE_VERSION_INFO)
+    if(NOT "${OUTPUT_VARIABLE}" MATCHES ".*GNU.*")
+      unset(MAKE_PRG)
+    endif()
+  endif()
+  if(NOT MAKE_PRG)
+    message(FATAL_ERROR "GNU Make is required to build the dependencies.")
+  else()
+    message(STATUS "Found GNU Make at ${MAKE_PRG}")
+  endif()
+endif()
+
 option(ENABLE_CCACHE "enable caching compiler output" ON)
 option(ENABLE_EXTRA_TESTS "pyston extra tests" OFF)
 option(ENABLE_GIL "threading use GIL" ON)
@@ -122,7 +139,8 @@ ExternalProject_Add(libjemalloc
                     DEPENDS gitsubmodules
                     UPDATE_COMMAND autoconf
                     CONFIGURE_COMMAND ${CMAKE_SOURCE_DIR}/build_deps/jemalloc/configure --prefix=${CMAKE_BINARY_DIR}/jemalloc --enable-autogen --enable-prof-libunwind
-                    INSTALL_COMMAND make install_bin install_lib
+                    BUILD_COMMAND ${MAKE_PRG}
+                    INSTALL_COMMAND ${MAKE_PRG} install_bin install_lib
                     LOG_UPDATE ON
                     LOG_CONFIGURE ON
                     LOG_BUILD ON
@@ -197,7 +215,7 @@ ExternalProject_Add(libunwind
                     DEPENDS libunwind_patched
                     UPDATE_COMMAND autoreconf -i
                     CONFIGURE_COMMAND ${CMAKE_SOURCE_DIR}/build_deps/libunwind/configure ${LIBUNWIND_DEBUG_CFLAGS} --prefix=${CMAKE_BINARY_DIR}/build_deps/libunwind --enable-shared=0 --disable-block-signals ${LIBUNWIND_CONSERVATIVE_CHECKS} ${LIBUNWIND_DEBUG} ${LIBUNWIND_DEBUG_FRAME}
-                    BUILD_COMMAND make -j${TEST_THREADS}
+                    BUILD_COMMAND ${MAKE_PRG} -j${TEST_THREADS}
                     LOG_UPDATE ON
                     LOG_CONFIGURE ON
                     LOG_BUILD ON

